src/cmd/flux-mini.py

```diff
@@ -0,0 +1,305 @@
+        }
+
+    def __create_resource(self, res_type, count, with_child=[]):
+        assert isinstance(
+            with_child, collectionsAbc.Sequence
+        ), "child resource must be a sequence"
+        assert not isinstance(with_child, str), "child resource must not be a string"
+        assert count > 0, "resource count must be > 0"
+
+        res = {"type": res_type, "count": count}
+
+        if len(with_child) > 0:
+            res["with"] = with_child
+        return res
+
+    def __create_slot(self, label, count, with_child):
+        slot = self.__create_resource("slot", count, with_child)
+        slot["label"] = label
+        return slot
+
+    def __parse_fsd(self, s):
+        m = re.match(r".*([smhd])$", s)
+        try:
+            n = float(s[:-1] if m else s)
+        except:
+            raise ValueError("invalid Flux standard duration")
+        unit = m.group(1) if m else "s"
+
+        if unit == "m":
+            seconds = timedelta(minutes=n).total_seconds()
+        elif unit == "h":
+            seconds = timedelta(hours=n).total_seconds()
+        elif unit == "d":
+            seconds = timedelta(days=n).total_seconds()
+        else:
+            seconds = n
+        if seconds < 0 or math.isnan(seconds) or math.isinf(seconds):
+            raise ValueError("invalid Flux standard duration")
+        return seconds
+
+    def set_duration(self, duration):
+        """
+        Assign a time limit to the job.  The duration may be:
+        - a float in seconds
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if type(duration) is str:
+            t = self.__parse_fsd(duration)
+        elif type(duration) is float:
+            t = duration
+        else:
+            raise ValueError("duration must be a float or string")
+        if t < 0:
+            raise ValueError("duration must not be negative")
+        if math.isnan(t) or math.isinf(t):
+            raise ValueError("duration must be a normal, finite value")
+        self.jobspec["attributes"]["system"]["duration"] = t
+
+    def set_cwd(self, cwd):
+        """
+        Set working directory of job.
+        """
+        if type(cwd) is not str:
+            raise ValueError("cwd must be a string")
+        self.jobspec["attributes"]["system"]["cwd"] = cwd
+
+    def set_environment(self, environ):
+        """
+        Set (entire) environment of job.
+        """
+        if type(environ) is not dict:
+            raise ValueError("environment must be a dictionary")
+        self.jobspec["attributes"]["system"]["environment"] = environ
+
+    def set_shell_options(self, opts):
+        """
+        Set (all) Flux shell options for job.
+        """
+        if type(opts) is not dict:
+            raise ValueError("shell options must be a dictionary")
+        self.jobspec["attributes"]["system"]["shell"]["options"] = opts
```

Comment:
This function isn't used yet, but my feeling is that we want to be able to specify `-o` multiple times on the command line with something like `-o mpi=spectrum -o cpu-affinity=off` (or equivalently a comma separated list). Does this function completely reassign all options? We might want to add a way to incrementally add them instead.

Reply:
Thanks!  Good feedback as I start working on that.